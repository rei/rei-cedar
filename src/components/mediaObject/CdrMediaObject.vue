<script setup lang="ts">
import { useCssModule, computed } from 'vue';
import mapClasses from '../../utils/mapClasses';
import { MediaObject, NameValuePair, HtmlAttributes } from '../../types/interfaces';
import type { Breakpoint } from '../../types/other';
import { modifyClassName } from '../../utils/buildClass';
import { getLayoutStyling } from '../../utils/mediaObject';
import CdrLayout from '../layout/CdrLayout.vue';
import { breakpoints, spacing } from '../../utils/other';

/** Component that pairs images or videos with related text in a consistent way */

defineOptions({ name: 'CdrMediaObject' });

const props = withDefaults(defineProps<MediaObject>(), {
  align: 'start',
  mediaPosition: 'left',
  mediaWidth: '1fr',
  mediaHeight: 'auto',
  mediaCover: false,
  overlay: false,
  overlayRowAlign: 'start',
  overlayColumnAlign: 'start',
  contentPadding: 'zero',
});

const style = useCssModule();

const rootProps = computed(() => {
  const baseClass = 'cdr-media-object';
  const classes = [baseClass];
  const inlineStyles: NameValuePair = {};
  const {
    align,
    mediaPosition,
    mediaWidth,
    mediaHeight,
    mediaCover,
    overlay,
    overlayRowAlign,
    overlayColumnAlign,
    contentPadding,
    ...otherProps
  } = props;
  const additionalProps: HtmlAttributes = { ...otherProps };

  if (contentPadding !== 'zero') {
    if (typeof contentPadding === 'string') {
      inlineStyles['--cdr-media-object-content-padding'] = spacing[contentPadding];
    } else {
      classes.push(modifyClassName(baseClass, 'content-padding-cq'));

      breakpoints.forEach((breakpoint: Breakpoint) => {
        // Add in padding styles for various breakpoints
        inlineStyles[`--cdr-media-object-content-padding-${breakpoint}`] =
          spacing[contentPadding[breakpoint]];
      });
    }
  }

  // Enter overlay mode, which uses media heights and widths as dimensions for
  // the entire container
  if (overlay) {
    classes.push(modifyClassName(baseClass, 'overlay'));
    Object.assign(additionalProps, { rows: 'auto', columns: 'auto' });
    inlineStyles['--cdr-media-object-row-align'] = overlayRowAlign;
    inlineStyles['--cdr-media-object-column-align'] = overlayColumnAlign;

    // Use mediaHeight to generate row height. Using 'auto' results in 0px height.
    if (mediaHeight === 'auto') {
      // Defaults to 9/16 ratio of container
      additionalProps.rows = '56cqw';
    } else {
      additionalProps.rows = mediaHeight;
    }
  } else {
    // Get layout related props and inline styles based on media measurements and
    // content position, both of which can be dynamic
    const layoutStyling = getLayoutStyling(mediaPosition, mediaWidth, mediaHeight);
    Object.assign(inlineStyles, layoutStyling.inlineStyles);
    Object.assign(additionalProps, layoutStyling.props);

    // Add in class for allowing dynamic content positioning
    if (typeof mediaPosition !== 'string') {
      classes.push(modifyClassName(baseClass, 'media-position-cq'));
    }

    // Add align class and inline styles for allowing dynamic align values
    // or set the static value
    if (typeof align !== 'string') {
      classes.push(modifyClassName(baseClass, 'align-cq'));

      breakpoints.forEach((breakpoint: Breakpoint) => {
        // Add in media position styles for various breakpoints
        inlineStyles[`--cdr-media-object-align-${breakpoint}`] = align[breakpoint];
      });
    } else {
      inlineStyles['--cdr-media-object-align'] = align;
    }

    // Add cover class
    if (mediaCover) {
      classes.push(modifyClassName(baseClass, 'cover'));
    }
  }

  return {
    ...additionalProps,
    class: mapClasses(style, ...classes) || undefined,
    style: inlineStyles,
  };
});
</script>

<template>
  <CdrLayout v-bind="rootProps">
    <div>
      <!-- @slot Where the media should be placed. Should be a single node. -->
      <slot name="media" />
    </div>
    <div>
      <!-- @slot Where all content should be placed. Can be multiple nodes. -->
      <slot name="content" />
    </div>
  </CdrLayout>
</template>

<style lang="scss" module src="./styles/CdrMediaObject.module.scss"></style>
