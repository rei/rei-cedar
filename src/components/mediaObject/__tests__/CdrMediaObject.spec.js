import { mount } from '../../../../test/vue-jest-style-workaround.js';
import CdrMediaObject from '../CdrMediaObject.vue';

describe('CdrMediaObject', () => {
  describe('snapshot test', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = mount(CdrMediaObject, { props: { checked: false } });
    });
    it('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });
});
