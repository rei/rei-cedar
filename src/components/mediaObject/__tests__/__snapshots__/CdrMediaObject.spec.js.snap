// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CdrMediaObject > snapshot tests > align center 1`] = `
<div
  class="cdr-layout cdr-layout--rows cdr-layout--columns cdr-media-object"
  style="--cdr-layout-rows: auto; --cdr-layout-columns: auto 1fr; --cdr-media-object-media-position: 'media content'; --cdr-media-object-align: center;"
>
  <!-- @slot Where all default content should be placed. -->
  
  <div
    class="cdr-media-object__media"
  >
    <!-- @slot Where the media should be placed. Should be a single node. -->
    
    Some media
    
  </div>
  <div
    class="cdr-media-object__content"
  >
    <!-- @slot Where all content should be placed. Can be multiple nodes. -->
    
    Some content
    
  </div>
  
</div>
`;

exports[`CdrMediaObject > snapshot tests > content padding 1`] = `
<div
  class="cdr-layout cdr-layout--rows cdr-layout--columns cdr-media-object"
  style="--cdr-layout-rows: auto; --cdr-layout-columns: 1fr 1fr; --cdr-media-object-content-padding: 32px; --cdr-media-object-media-position: 'media content'; --cdr-media-object-align: start;"
>
  <!-- @slot Where all default content should be placed. -->
  
  <div
    class="cdr-media-object__media"
  >
    <!-- @slot Where the media should be placed. Should be a single node. -->
    
    Some media
    
  </div>
  <div
    class="cdr-media-object__content"
  >
    <!-- @slot Where all content should be placed. Can be multiple nodes. -->
    
    Some content
    
  </div>
  
</div>
`;

exports[`CdrMediaObject > snapshot tests > content padding dynamic 1`] = `
<div
  class="cdr-layout cdr-layout--rows cdr-layout--columns cdr-media-object cdr-media-object--content-padding-cq"
  style="--cdr-layout-rows: auto; --cdr-layout-columns: 1fr 1fr; --cdr-media-object-content-padding-xs: 16px; --cdr-media-object-content-padding-sm: 16px; --cdr-media-object-content-padding-md: 48px; --cdr-media-object-content-padding-lg: 48px; --cdr-media-object-media-position: 'media content'; --cdr-media-object-align: start;"
>
  <!-- @slot Where all default content should be placed. -->
  
  <div
    class="cdr-media-object__media"
  >
    <!-- @slot Where the media should be placed. Should be a single node. -->
    
    Some media
    
  </div>
  <div
    class="cdr-media-object__content"
  >
    <!-- @slot Where all content should be placed. Can be multiple nodes. -->
    
    Some content
    
  </div>
  
</div>
`;

exports[`CdrMediaObject > snapshot tests > cover 1`] = `
<div
  class="cdr-layout cdr-layout--rows cdr-layout--columns cdr-media-object cdr-media-object--cover"
  style="--cdr-layout-rows: auto; --cdr-layout-columns: 125px 1fr; --cdr-media-object-media-position: 'media content'; --cdr-media-object-align: start;"
>
  <!-- @slot Where all default content should be placed. -->
  
  <div
    class="cdr-media-object__media"
  >
    <!-- @slot Where the media should be placed. Should be a single node. -->
    
    Some media
    
  </div>
  <div
    class="cdr-media-object__content"
  >
    <!-- @slot Where all content should be placed. Can be multiple nodes. -->
    
    Some content
    
  </div>
  
</div>
`;

exports[`CdrMediaObject > snapshot tests > default: media position left, media width 1fr, media height auto, align start 1`] = `
<div
  class="cdr-layout cdr-layout--rows cdr-layout--columns cdr-media-object"
  style="--cdr-layout-rows: auto; --cdr-layout-columns: 1fr 1fr; --cdr-media-object-media-position: 'media content'; --cdr-media-object-align: start;"
>
  <!-- @slot Where all default content should be placed. -->
  
  <div
    class="cdr-media-object__media"
  >
    <!-- @slot Where the media should be placed. Should be a single node. -->
    
    Some media
    
  </div>
  <div
    class="cdr-media-object__content"
  >
    <!-- @slot Where all content should be placed. Can be multiple nodes. -->
    
    Some content
    
  </div>
  
</div>
`;

exports[`CdrMediaObject > snapshot tests > dynamic align 1`] = `
<div
  class="cdr-layout cdr-layout--rows cdr-layout--columns cdr-media-object cdr-media-object--align-cq"
  style="--cdr-layout-rows: auto; --cdr-layout-columns: auto 1fr; --cdr-media-object-media-position: 'media content'; --cdr-media-object-align-xs: center; --cdr-media-object-align-sm: center; --cdr-media-object-align-md: end; --cdr-media-object-align-lg: start;"
>
  <!-- @slot Where all default content should be placed. -->
  
  <div
    class="cdr-media-object__media"
  >
    <!-- @slot Where the media should be placed. Should be a single node. -->
    
    Some media
    
  </div>
  <div
    class="cdr-media-object__content"
  >
    <!-- @slot Where all content should be placed. Can be multiple nodes. -->
    
    Some content
    
  </div>
  
</div>
`;

exports[`CdrMediaObject > snapshot tests > media bottom 1`] = `
<div
  class="cdr-layout cdr-layout--rows cdr-layout--columns cdr-media-object"
  style="--cdr-layout-rows: 1fr auto; --cdr-layout-columns: 1fr; --cdr-media-object-media-position: 'content' 'media'; --cdr-media-object-align: start;"
>
  <!-- @slot Where all default content should be placed. -->
  
  <div
    class="cdr-media-object__media"
  >
    <!-- @slot Where the media should be placed. Should be a single node. -->
    
    Some media
    
  </div>
  <div
    class="cdr-media-object__content"
  >
    <!-- @slot Where all content should be placed. Can be multiple nodes. -->
    
    Some content
    
  </div>
  
</div>
`;

exports[`CdrMediaObject > snapshot tests > media position dynamic 1`] = `
<div
  class="cdr-layout cdr-layout--rows cdr-layout--rows-cq cdr-layout--columns cdr-layout--columns-cq cdr-media-object cdr-media-object--media-position-cq"
  style="--cdr-layout-rows-xs: auto 1fr; --cdr-layout-rows-sm: 1fr auto; --cdr-layout-rows-md: auto; --cdr-layout-rows-lg: auto; --cdr-layout-columns-xs: 1fr; --cdr-layout-columns-sm: 1fr; --cdr-layout-columns-md: 1fr 1fr; --cdr-layout-columns-lg: 1fr 1fr; --cdr-media-object-media-position-xs: 'media' 'content'; --cdr-media-object-media-position-sm: 'content' 'media'; --cdr-media-object-media-position-md: 'media content'; --cdr-media-object-media-position-lg: 'content media'; --cdr-media-object-align: start;"
>
  <!-- @slot Where all default content should be placed. -->
  
  <div
    class="cdr-media-object__media"
  >
    <!-- @slot Where the media should be placed. Should be a single node. -->
    
    Some media
    
  </div>
  <div
    class="cdr-media-object__content"
  >
    <!-- @slot Where all content should be placed. Can be multiple nodes. -->
    
    Some content
    
  </div>
  
</div>
`;

exports[`CdrMediaObject > snapshot tests > media position dynamic, media height dynamic, media width dynamic 1`] = `
<div
  class="cdr-layout cdr-layout--rows cdr-layout--rows-cq cdr-layout--columns cdr-layout--columns-cq cdr-media-object cdr-media-object--media-position-cq cdr-media-object--cover"
  style="--cdr-layout-rows-xs: 100px 1fr; --cdr-layout-rows-sm: 200px 1fr; --cdr-layout-rows-md: auto; --cdr-layout-rows-lg: auto; --cdr-layout-columns-xs: 100%; --cdr-layout-columns-sm: 100%; --cdr-layout-columns-md: 50% 1fr; --cdr-layout-columns-lg: 75% 1fr; --cdr-media-object-media-position-xs: 'media' 'content'; --cdr-media-object-media-position-sm: 'media' 'content'; --cdr-media-object-media-position-md: 'media content'; --cdr-media-object-media-position-lg: 'media content'; --cdr-media-object-align: start;"
>
  <!-- @slot Where all default content should be placed. -->
  
  <div
    class="cdr-media-object__media"
  >
    <!-- @slot Where the media should be placed. Should be a single node. -->
    
    Some media
    
  </div>
  <div
    class="cdr-media-object__content"
  >
    <!-- @slot Where all content should be placed. Can be multiple nodes. -->
    
    Some content
    
  </div>
  
</div>
`;

exports[`CdrMediaObject > snapshot tests > media position right 1`] = `
<div
  class="cdr-layout cdr-layout--rows cdr-layout--columns cdr-media-object"
  style="--cdr-layout-rows: auto; --cdr-layout-columns: 1fr 1fr; --cdr-media-object-media-position: 'content media'; --cdr-media-object-align: start;"
>
  <!-- @slot Where all default content should be placed. -->
  
  <div
    class="cdr-media-object__media"
  >
    <!-- @slot Where the media should be placed. Should be a single node. -->
    
    Some media
    
  </div>
  <div
    class="cdr-media-object__content"
  >
    <!-- @slot Where all content should be placed. Can be multiple nodes. -->
    
    Some content
    
  </div>
  
</div>
`;

exports[`CdrMediaObject > snapshot tests > media top, align center 1`] = `
<div
  class="cdr-layout cdr-layout--rows cdr-layout--columns cdr-media-object"
  style="--cdr-layout-rows: auto 1fr; --cdr-layout-columns: 1fr; --cdr-media-object-media-position: 'media' 'content'; --cdr-media-object-align: center;"
>
  <!-- @slot Where all default content should be placed. -->
  
  <div
    class="cdr-media-object__media"
  >
    <!-- @slot Where the media should be placed. Should be a single node. -->
    
    Some media
    
  </div>
  <div
    class="cdr-media-object__content"
  >
    <!-- @slot Where all content should be placed. Can be multiple nodes. -->
    
    Some content
    
  </div>
  
</div>
`;

exports[`CdrMediaObject > snapshot tests > media width 100px 1`] = `
<div
  class="cdr-layout cdr-layout--rows cdr-layout--columns cdr-media-object"
  style="--cdr-layout-rows: auto; --cdr-layout-columns: 100px 1fr; --cdr-media-object-media-position: 'media content'; --cdr-media-object-align: start;"
>
  <!-- @slot Where all default content should be placed. -->
  
  <div
    class="cdr-media-object__media"
  >
    <!-- @slot Where the media should be placed. Should be a single node. -->
    
    Some media
    
  </div>
  <div
    class="cdr-media-object__content"
  >
    <!-- @slot Where all content should be placed. Can be multiple nodes. -->
    
    Some content
    
  </div>
  
</div>
`;

exports[`CdrMediaObject > snapshot tests > media width dynamic 1`] = `
<div
  class="cdr-layout cdr-layout--rows cdr-layout--columns cdr-layout--columns-cq cdr-media-object"
  style="--cdr-layout-rows: auto; --cdr-layout-columns-xs: 100px 1fr; --cdr-layout-columns-sm: 100px 1fr; --cdr-layout-columns-md: 300px 1fr; --cdr-layout-columns-lg: 300px 1fr; --cdr-media-object-media-position: 'media content'; --cdr-media-object-align: start;"
>
  <!-- @slot Where all default content should be placed. -->
  
  <div
    class="cdr-media-object__media"
  >
    <!-- @slot Where the media should be placed. Should be a single node. -->
    
    Some media
    
  </div>
  <div
    class="cdr-media-object__content"
  >
    <!-- @slot Where all content should be placed. Can be multiple nodes. -->
    
    Some content
    
  </div>
  
</div>
`;

exports[`CdrMediaObject > snapshot tests > overlay, row align, column align, content configured independently to 50% width 1`] = `
<div
  class="cdr-layout cdr-layout--rows cdr-layout--columns cdr-media-object cdr-media-object--overlay"
  style="--cdr-layout-rows: auto; --cdr-layout-columns: auto; --cdr-media-object-row-align: center; --cdr-media-object-column-align: end;"
>
  <!-- @slot Where all default content should be placed. -->
  
  <div
    class="cdr-media-object__media"
  >
    <!-- @slot Where the media should be placed. Should be a single node. -->
    
    Some media
    
  </div>
  <div
    class="cdr-media-object__content"
  >
    <!-- @slot Where all content should be placed. Can be multiple nodes. -->
    
    Some content
    
  </div>
  
</div>
`;

exports[`CdrMediaObject > snapshot tests > pass down props to Layout and Surface 1`] = `
<div
  class="cdr-surface cdr-surface--modifier-default cdr-surface--background-brand-spruce cdr-surface--radius-sharp cdr-surface--shadow-flat cdr-layout cdr-layout--rows cdr-layout--columns cdr-media-object"
  data-palette="default"
  style="--cdr-layout-rows: auto; --cdr-layout-columns: 1fr 1fr; --cdr-media-object-media-position: 'media content'; --cdr-media-object-align: start;"
>
  <!-- @slot Where all default content should be placed. -->
  
  <!-- @slot Where all default content should be placed. -->
  
  <div
    class="cdr-media-object__media"
  >
    <!-- @slot Where the media should be placed. Should be a single node. -->
    
    Some media
    
  </div>
  <div
    class="cdr-media-object__content"
  >
    <!-- @slot Where all content should be placed. Can be multiple nodes. -->
    
    Some content
    
  </div>
  
  
</div>
`;
